Decouple peer-processing-logic from block-connection-logic

TheBlueMatt opened the pull request:
This is part of a series of about 20 commits which split main.cpp into two - peer processing logic and blockchain/mempool/UTXO logic. Most of them arent much more intrusive than these, at least until there are large blocks of code moving.

This set focuses primarily on using CValidationinterface to handle peer-processing-logic updates when connecting/disconnecting blocks.

I haven't significantly tested this as my normal test machine is largely unavailable atm, but most of the changes here are pretty straight-forward.

TheBlueMatt committed:
Make validationinterface.UpdatedBlockTip more verbose

In anticipation of making all the callbacks out of block processing
flow through it. Note that vHashes will always have something in it
since pindexFork != pindexNewTip.

TheBlueMatt committed:
Remove duplicate nBlocksEstimate cmp (we already checked IsIBD())

TheBlueMatt committed:
Remove CConnman parameter from ProcessNewBlock/ActivateBestChain

TheBlueMatt committed:
Remove SyncWithWallets wrapper function

theuni commented on src/main.cpp:
```diff
@@ -4684,6 +4658,62 @@ std::string GetWarnings(const std::string& strFor)
+void InitPeerLogic(CConnman& connman) {
+    if (mapPeerLogics.count(&connman))
+        return;
+    mapPeerLogics.emplace(std::make_pair(&connman, std::unique_ptr<PeerLogicValidation>(new PeerLogicValidation(&connman))));
```
nit: don't make_pair with emplace, just use the components directly.

theuni commented on src/main.cpp:
```diff
@@ -4684,6 +4658,62 @@ std::string GetWarnings(const std::string& strFor)
+        }
+    }
+};
+std::map<CConnman*, std::unique_ptr<PeerLogicValidation> > mapPeerLogics;
```
Using a ptr for the index feels kinda icky. Since a reference is passed and stored anyway, let's add an Id to CConnman instead. It'll be helpful in other places too.

TheBlueMatt replied:
Hmm, I feel like that is a layer violation? If CConnman is a low-level network socket/connection manager, it should have little knowledge of the fact that there is something hooked up above it to listen to validation callbacks and do things with them?

theuni replied:
Sorry, I meant: add a GetID() to CConnman, which would return a const value set during construction. That would allow maps to be created which refer to specific connman instances without tying them to a pointer.

theuni replied:
I'm hoping to be able to hook up 2 CConnmans to eachother in the future for testing against eachother, yes. I've gone to significant efforts to remove the "stuff a static global somewhere" practice from much of the net code, please don't start reintroducing.

A pointer is fine here if you'd prefer to keep it simple, we can adapt later if needed.

theuni commented on src/main.cpp:
```diff
@@ -4693,6 +4681,26 @@ class PeerLogicValidation : public CValidationInterface {
+            }
+        }
+        if (it != mapBlockSource.end())
+            mapBlockSource.erase(it);
```
Hmm, were these not erased before in the invalid case?

theuni commented:
Concept ACK and quick code-review ACK other than the nits. I'd like to spend some time testing, though.

TheBlueMatt commented:
No? I think after this patch set the only place mapBlockSource is erased from is this line.

On October 3, 2016 8:26:53 PM EDT, Cory Fields <notifications@github.com> wrote:
>theuni commented on this pull request.
>
>
>
>> +
>+        const uint256 hash(block.GetHash());
>+        std::map<uint256, NodeId>::iterator it =
>mapBlockSource.find(hash);
>+
>+        int nDoS = 0;
>+        if (state.IsInvalid(nDoS)) {
>+            if (it != mapBlockSource.end() && State(it->second)) {
>+                assert (state.GetRejectCode() < REJECT_INTERNAL); //
>Blocks are never rejected with internal reject codes
>+                CBlockReject reject = {(unsigned
>char)state.GetRejectCode(), state.GetRejectReason().substr(0,
>MAX_REJECT_MESSAGE_LENGTH), hash};
>+                State(it->second)->rejects.push_back(reject);
>+                if (nDoS > 0)
>+                    Misbehaving(it->second, nDoS);
>+            }
>+        }
>+        if (it != mapBlockSource.end())
>+            mapBlockSource.erase(it);
>
>Hmm, were these not erased before in the invalid case?
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/8865#pullrequestreview-2631559

theuni commented:
I was just pointing out that this looks to be a memleak bugfix :)

TheBlueMatt commented:
I dont believe it is actually any (much?) different than it used to be. Maybe it will be deleted more often on reorgs now, I didnt investigate fully. In any case, its pretty hard to DoS since you only add to the map after AcceptBlock succeeds (ie the block is syntactically valid/connects/has valid PoW)

TheBlueMatt commented:
Switched g_connman to be a shared_ptr instead of a unique_ptr, which I think should neatly address all the outstanding nits.

TheBlueMatt commented:
Backed out the shared_ptr change, I misunderstood @theuni's comments, just fixed the make_pair nit and left everything else as-is: it should be easier to clean up things like class design once everything is split out.

TheBlueMatt committed:
Move net-processing logic definitions together in main.h

TheBlueMatt committed:
Use CValidationInterface from chain logic to notify peer logic

This adds a new CValidationInterface subclass, defined in main.h,
to receive notifications of UpdatedBlockTip and use that to push
blocks to peers, instead of doing it directly from
ActivateBestChain.

TheBlueMatt committed:
Always call UpdatedBlockTip, even if blocks were only disconnected

TheBlueMatt committed:
Use BlockChecked signal to send reject messages from mapBlockSource

TheBlueMatt commented:
Restructured the code after discussion on IRC with @theuni...now the CValidationInterface is defined in main.h (requires an extra include, but we can drop that when we move all the stuff to another file) and init.cpp itself does the object initialization and keeps the CValidationInterface around.

laanwj commented:
utACK a9aec5c

theuni commented on src/main.cpp:
```diff
@@ -3087,12 +3087,9 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
-                    connman->ForEachNode([nNewHeight, nBlockEstimate, &vHashes](CNode* pnode) {
-                        if (nNewHeight > (pnode->nStartingHeight != -1 ? pnode->nStartingHeight - 2000 : nBlockEstimate)) {
+                    connman->ForEachNode([nNewHeight, &vHashes](CNode* pnode) {
+                        if (nNewHeight > (pnode->nStartingHeight != -1 ? pnode->nStartingHeight - 2000 : 0)) {
```
If pnode->nStartingHeight is -1, we just haven't completed the version handshake, no? Shouldn't we always refuse to relay in that case?

Or am I not seeing some other way this can happen?

TheBlueMatt replied:
Indeed, probably, though that would be a behavior change whereas this PR shouldnt have any :p.

theuni commented on src/validationinterface.h:
```diff
@@ -50,6 +48,8 @@ class CValidationInterface {
     /** Notifies listeners of updated block chain tip */
     boost::signals2::signal<void (const CBlockIndex *, const CBlockIndex *, bool fInitialDownload)> UpdatedBlockTip;
+    /** A posInBlock value for SyncTransaction which indicates the transaction was conflicted, disconnected, or not in a block */
+    static const int SYNC_TRANSACTION_NOT_IN_BLOCK = -1;
```
This is fine for now, but I think we ultimately want these split up.

TheBlueMatt replied:
Yea, I think medium-term we need to think hard about what goes in what signaling interface (we have two already that are highly duplicative, I have some thoughts but we should discuss)

theuni commented:
post-merge utACK.

rebroad commented on src/main.cpp:
```diff
@@ -4689,6 +4644,59 @@ std::string GetWarnings(const std::string& strFor)
+        }
+        // Relay inventory, but don't relay old inventory during initial block download.
+        connman->ForEachNode([nNewHeight, &vHashes](CNode* pnode) {
+            if (nNewHeight > (pnode->nStartingHeight != -1 ? pnode->nStartingHeight - 2000 : 0)) {
```
The logic of this line is confusing now... If pnode->nStartingHeight = -1 then don't do (if nNewHeight > -1) but instead do (if nNewHeight > 0)?! It hardly makes any difference, it will return true, therefore simply do if (nNewHeight > pnode->nStartingHeight) without any need for the ? operator.

Anyway, why make it always true and not make use of the nBlocksEstimate as was previously done?

rebroad replied:
Fixed in #8958

sipa replied:
See the earlier commit: "Remove duplicate nBlocksEstimate cmp (we already checked IsIBD())".

sipa commented:
Posthummus utACK.
